eth/handler_bor.go

```diff
@@ -39,80 +39,148 @@ var (
+	errNoMilestone = errors.New("no milestone proposed")
+
+	// errMilestone is returned when we are unable to fetch the
+	// latest checkpoint from the local heimdall.
+	errMilestone = errors.New("failed to fetch the milestone")
+
+	// errMissingMilestone is returned when we don't have the
+	// checkpoint blocks locally, yet.
+	errMissingMilestone = errors.New("missing checkpoint blocks")
+
+	// errMilestoneRootHashMismatch is returned when the local root hash
+	// doesn't match with the root hash in milestone.
+	errMilestoneRootHashMismatch = errors.New("milestone roothash mismatch")
 )
 
 // fetchWhitelistCheckpoints fetches the latest checkpoint/s from it's local heimdall
 // and verifies the data against bor data.
-func (h *ethHandler) fetchWhitelistCheckpoints(ctx context.Context, bor *bor.Bor, first bool) ([]uint64, []common.Hash, error) {
+func (h *ethHandler) fetchWhitelistCheckpoint(ctx context.Context, bor *bor.Bor, first bool) (uint64, common.Hash, error) {
 	// Create an array for block number and block hashes
 	//nolint:prealloc
 	var (
-		blockNums   []uint64      = make([]uint64, 0)
-		blockHashes []common.Hash = make([]common.Hash, 0)
+		blockNum  uint64
+		blockHash common.Hash
 	)
 
 	// Fetch the checkpoint count from heimdall
 	count, err := bor.HeimdallClient.FetchCheckpointCount(ctx)
 	if err != nil {
 		log.Debug("Failed to fetch checkpoint count for whitelisting", "err", err)
-		return blockNums, blockHashes, errCheckpointCount
+		return blockNum, blockHash, errCheckpointCount
+	}
+
+	if count == 0 {
+		return blockNum, blockHash, errNoCheckpoint
+	}
+
+	// fetch the latest checkpoint from Heimdall
+	checkpoint, err := bor.HeimdallClient.FetchCheckpoint(ctx)
+	if err != nil {
+		log.Debug("Failed to fetch latest checkpoint for whitelisting", "err", err)
+		return blockNum, blockHash, errMilestone
+	}
+
+	// check if we have the checkpoint blocks
+	head := h.ethAPI.BlockNumber()
+	if head < hexutil.Uint64(checkpoint.EndBlock.Uint64()) {
+		log.Debug("Head block behind checkpoint block", "head", head, "milestone end block", checkpoint.EndBlock)
+		return blockNum, blockHash, errMissingMilestone
+	}
+
+	// verify the root hash of checkpoint
+	roothash, err := h.ethAPI.GetRootHash(ctx, checkpoint.StartBlock.Uint64(), checkpoint.EndBlock.Uint64())
+	if err != nil {
+		log.Debug("Failed to get root hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errRootHash
+	}
+
+	if roothash != checkpoint.RootHash.String()[2:] {
+		log.Warn("Checkpoint root hash mismatch while whitelisting", "expected", checkpoint.RootHash.String()[2:], "got", roothash)
+		return blockNum, blockHash, errCheckpointRootHashMismatch
+	}
+
+	// fetch the end checkpoint block hash
+	block, err := h.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(checkpoint.EndBlock.Uint64()), false)
+	if err != nil {
+		log.Debug("Failed to get end block hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errEndBlock
```

Comment:
why don't what an error and remove logging on that level?

Reply:
Do you mean we should return the err itself rather than creating errEndBlock

Reply:
Like add the context to the error with fmt Errorf + %w